Fixes phpcs  errors and warnings for Magento\Framework\Image

krisdante opened the pull request:
### Description (*)

I was working with those classes and spotted a few warnings and errors
that prevent smooth pull requests. This PR fixes them.
It introduces Localized Exceptions in place of \Exception too.

<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->


<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1. `vendor/bin/phpcs --standard=dev/tests/static/framework/Magento/ruleset.xml lib/internal/Magento/Framework/Image/`
### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

krisdante committed:
Fixes phpcs  errors and warnings for Magento\Framework\Image

I was working with those classes and spotted a few warnings and errors
that prevent from smooth pull requests. This PR fixes them.

It introduces Localized Exceptions in place of \Exception too.

Watermark method was to long so it was devided to separate parts

lbajsarowicz commented on lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php:
```diff
@@ -66,7 +71,7 @@ public function backgroundColor($color = null)
      * @param string $filename
      * @return void
-     * @throws \Exception
+     * @throws \Magento\Framework\Exception\LocalizedException
```
Use imported class name

krisdante replied:
Let's discuss this in one place

lbajsarowicz commented on lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php:
```diff
@@ -77,7 +82,11 @@ public function open($filename)
         } catch (\ImagickException $e) {
-            throw new \Exception(sprintf('Unsupported image format. File: %s', $this->_fileName), $e->getCode(), $e);
+            throw new LocalizedException(
+                new \Magento\Framework\Phrase(sprintf('Unsupported image format. File: %s', $this->_fileName)),
```
Whole line can be replaced with `__('Unsupported image format. File: %1', $this->_fileName),`

lbajsarowicz commented on lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php:
```diff
@@ -86,12 +95,13 @@ public function open($filename)
      * @param null|string $newName
      * @return void
-     * @throws \Exception  If destination path is not writable
+     * @throws \Magento\Framework\Exception\LocalizedException If destination path is not writable
```
Use imported class name

lbajsarowicz commented on lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php:
```diff
@@ -241,7 +253,7 @@ public function crop($top = 0, $left = 0, $right = 0, $bottom = 0)
      * @return void
      * @throws \LogicException
-     * @throws \Exception
+     * @throws \Magento\Framework\Exception\LocalizedException
```
Use imported class name

krisdante replied:
But why? Extracted documentation is not aware of the imports. 
I see those everywhere in Magento Code:
```
    /**
     * Remove product from Bundle product option
     *
     * @param string $sku
     * @param int $optionId
     * @param string $childSku
     * @throws \Magento\Framework\Exception\NoSuchEntityException
     * @throws \Magento\Framework\Exception\InputException
     * @return bool
     */
    public function removeChild($sku, $optionId, $childSku);
```

lbajsarowicz replied:
Let me discuss that. Let's keep it like this for now :-)

lbajsarowicz commented on lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php:
```diff
@@ -333,7 +345,11 @@ public function watermark($imagePath, $positionX = 0, $positionY = 0, $opacity =
         } catch (\ImagickException $e) {
-            throw new \Exception('Unable to create watermark.', $e->getCode(), $e);
+            throw new LocalizedException(
+                new \Magento\Framework\Phrase('Unable to create watermark.'),
```
`__('Unable to create watermark.')`

krisdante replied:
I think the whole idea behind LocalizedException and Phrase is not to use translation function on top of it.

krisdante replied:
Nowhere in code it is used this way

lbajsarowicz replied:
`__()` actually is new Phrase - take a look into https://github.com/magento/magento2/blob/c7c0a992d83de56f15555d3edcecef7aaa32f0d7/app/functions.php#L19:29

krisdante replied:
So using new Phrase is cleaner (OOP not superglobal function)

lbajsarowicz replied:
Both ways are correct unless you use full path for all `Phrase`s :-)

krisdante replied:
OK i checked other places in Magento. Changed

lbajsarowicz commented on lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php:
```diff
@@ -346,12 +362,16 @@ public function watermark($imagePath, $positionX = 0, $positionY = 0, $opacity =
      *
      * @return void
-     * @throws \Exception If some of dependencies are missing
+     * @throws \Magento\Framework\Exception\LocalizedException If some of dependencies are missing
```
Use imported class.

lbajsarowicz replied:
Mentioned previously.

lbajsarowicz commented on lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php:
```diff
@@ -346,12 +362,16 @@ public function watermark($imagePath, $positionX = 0, $positionY = 0, $opacity =
         if (!class_exists('\Imagick', false)) {
-            throw new \Exception("Required PHP extension 'Imagick' was not loaded.");
+            throw new LocalizedException(
+                new \Magento\Framework\Phrase('Required PHP extension \'Imagick\' was not loaded.'),
```
`__('Required PHP extension "Imagick" was not loaded.')`

lbajsarowicz commented on lib/internal/Magento/Framework/Image/Factory.php:
```diff
@@ -8,6 +8,10 @@
 
+/**
+ * Factory for object to manipulate images
+ *
```
Empty line?

lbajsarowicz commented on lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php:
```diff
@@ -241,7 +253,7 @@ public function crop($top = 0, $left = 0, $right = 0, $bottom = 0)
      * @param bool $tile
      * @return void
      * @throws \LogicException
```
Import class.

lbajsarowicz commented on lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php:
```diff
@@ -5,6 +5,11 @@
+/**
+ * Image adapter from ImageMagick
+ */
 class ImageMagick extends \Magento\Framework\Image\Adapter\AbstractAdapter
```
~Import class.~

Actually it's in the same namespace, so you can directly use AbstractAdapter without full path.

lbajsarowicz commented on lib/internal/Magento/Framework/Image/Adapter/Config.php:
```diff
@@ -18,13 +18,13 @@ class Config implements ConfigInterface, UploadConfigInterface
     /**
      * Config path for the maximal image height value
-     * @deprecated
+     * @deprecated Used in a method that is deprecate()
```
What is `deprecate()`?

krisdante replied:
Typo. will fix

lbajsarowicz commented on lib/internal/Magento/Framework/Image/Adapter/AbstractAdapter.php:
```diff
@@ -25,24 +25,42 @@ abstract class AbstractAdapter implements AdapterInterface
+    /**
+     * Position in the center
+     */
     const POSITION_CENTER = 'center';
```
You may still use single docblock for multiple constants:
```php
    /**#@+
     * Position constants
     */
    public const POSITION_TOP_LEFT = 'top-left';
    ...
    public const POSITION_CENTER = 'center';
    /**#@-*/
```

lbajsarowicz commented on lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php:
```diff
@@ -499,4 +480,84 @@ protected function _getImagickPixelObject($color = null)
+     *
+     * @param $positionX
+     * @param $positionY
+     * @param \Imagick $watermark
```
I'd import the `\Imagick` class.

krisdante replied:
What is the benefit of such import?

lbajsarowicz replied:
The imports are consistent.

krisdante replied:
I followed this suggestion and phpstan raged. Reversing

lbajsarowicz commented:
@krisdante Fix the Static tests and import `Imagick` class, then I'll approve PR :+1:

rodde177 commented on lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php:
```diff
@@ -499,4 +480,84 @@ protected function _getImagickPixelObject($color = null)
+     *
+     * @param \Imagick $watermark
+     */
+    private function resizeWatermerk(\Imagick $watermark): void
```
typo > `resizeWatermerk`

lbajsarowicz commented:
@magento run tests

krisdante commented:
Jezus this is boring.... what now. Why the way to test is locally is not documented. phpcs is passing for me.

krisdante commented:
```
/usr/bin/php /var/www/magentosrc/./vendor/phpunit/phpunit/phpunit -v --testdox testsuite/Magento/Test/Php/LiveCodeTest.php
PHPUnit 6.5.14 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.1.27 with Xdebug 2.7.2
Configuration: /var/www/creativeshop/magentosrc/dev/tests/static/phpunit.xml.dist

Magento\Test\Php\LiveCode
 [x] Code style
 [x] Code mess
 [x] Copy paste
 [x] Strict types
 [x] Php compatibility
 [x] Php stan
```

Those tests are passing with my clone of Magento. I do not know why they are failing for B2B test here. I cannot reproduce it. Also, the failed unit test has nothing to do with the change here

lbajsarowicz commented:
@krisdante Don't worry about Functional Tests - this is known bug of Magento `2.4-develop` branch and will be merged soon to `2.4-develop` (then you'll be able to rebase your branch onto it). Unless - just fix static tests.or optionally restart them if you believe that your code is fine.

Use
```
@magento run Static Tests
```

krisdante commented:
I do not know how to fix them. They are passing on local. See my message.

lbajsarowicz commented:
@krisdante Let me ping the right person for that kind of issue

lenaorobei commented:
This issue is caused because `ImageMagick` adapter is non default one and can be configured. 

Please add `lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php` to the `dev/tests/static/testsuite/Magento/Test/Php/_files/phpstan/blacklist/common.txt`.

lbajsarowicz commented:
@krisdante Could you make the changes requested by @lenaorobei, please?

krisdante committed:
Adds ImageMagick.php to blacklist of phpstan

ImageMagick adapter is non default one and can be configured
It is not included in Test Enviroment

krisdante commented:
@lbajsarowicz I did. Static tests are now passing. Failing B2B test is irrelevant.

lbajsarowicz reviewed (approved):
Thank you for the great contribution!

magento-engcom-team commented:
Hi @lbajsarowicz, thank you for the review.
[ENGCOM-7260](https://jira.corp.magento.com/browse/ENGCOM-7260) has been created to process this Pull Request
:eight_spoked_asterisk: @lbajsarowicz, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

krisdante commented:
This DB error is false. I am not going to re-run this test anymore. Merge it or delete the whole PR.

lbajsarowicz commented:
@krisdante Don't worry about false negatives. These are also verified manually by Commeng team, so as long as it's false negative, we're able to skip it.

engcom-Alfa commented:
:heavy_check_mark: QA Passed

vendor/bin/phpcs --standard=dev/tests/static/framework/Magento/ruleset.xml lib/internal/Magento/Framework/Image/

**Before:**
![Screenshot from 2020-04-07 14-27-30](https://user-images.githubusercontent.com/51679138/78667512-c1a4d000-78e1-11ea-9444-c0e3c5e395b3.png)

**After:**
![phpcs](https://user-images.githubusercontent.com/51679138/78667526-c8334780-78e1-11ea-8885-724e7aacdcff.png)

engcom-Echo commented:
Failed Database Compare not related to the changes in this PR
